Sources/CSRF.swift

```diff
@@ -1,53 +1,46 @@
-import Foundation
-import Vapor
-import HTTP
-import Node
-import JSON
 import Crypto
+import Vapor
 
-public typealias TokenRetrievalHandler = ((Request) throws -> String)!
+public typealias TokenRetrievalHandler = ((Request) throws -> Future<String>)
 
 /// Middleware to protect against cross-site request forgery attacks.
-public struct CSRF: Middleware {
-    private let ignoredMethods: HTTPMethod
+public struct CSRF: Middleware, Service {
+    private let ignoredMethods: [HTTPMethod]
     private var tokenRetrieval: TokenRetrievalHandler
-    private let hasher = CryptoHasher(hash: .md5, encoding: .hex)
+    private let hasher = CryptoHasher()
```

Comment:
Why not call down to the MD5 directly instead of wrapping it in an inner struct? Also if people don't add CSRF correctly this could cause issues with thread-safety

Reply:
As you say, this codes is redundant, so I will fix.
What is thread safe issues? Is it fixed by using MD5 directly?

Reply:
The _potential_ thread safe issues are because you have a instance property `hasher` that if the middleware is registered globally (i.e. the same middleware for all threads) you could have multiple threads all using `hasher` at the same time. This _shouldn't_ cause an issue at the moment since it doesn't write anything, but I'm not sure if the hashing functions are thread-safe. Using MD5 directly (i.e. where it is needed as opposed to a class variable) should be fine since each request will create a new hasher

Reply:
I understand. I will fix it to be thread safety.